msrestazure/azure_operation.py

```diff
@@ -182,122 +200,59 @@ def _get_body_status(self, response):
-        Just sets status to 'InProgress'.
-
-        :param requests.Response response: latest REST call response.
-        """
-        self.status = 'InProgress'
+        return body.get("properties", {}).get("provisioningState")
 
-    def _status_204(self, response):
-        """Process response with status code 204.
-        Interpretted as successful with no payload.
+    def should_do_final_get(self):
+        """Check whether the polling should end doing a final GET.
 
         :param requests.Response response: latest REST call response.
-        """
-        self.status = 'Succeeded'
-        self.resource = None
-
-    def is_done(self):
-        """Check whether the operation can be considered complete.
-        This is based on whether the data in the resource matches the current
-        status. If there is not resource, we assume it's complete.
-
         :rtype: bool
         """
-        if (self.async_url or not self.resource) and \
-                self.method in ['PUT', 'PATCH']:
-            return False
-        resource_state = self._get_resource_status()
-        try:
-            return self.status.lower() == resource_state.lower()
-        except AttributeError:
-            return True
+        return (self.async_url or not self.resource) and \
+                self.method in {'PUT', 'PATCH'}
 
-    def get_initial_status(self, response):
+    def set_initial_status(self, response):
         """Process first response after initiating long running
-        operation.
+        operation and set self.status attribute.
 
         :param requests.Response response: initial REST call response.
         """
-        self._deserialize(response)
-        self._check_status(response)
-        if response.status_code in [200, 202, 204]:
-            self._process_status(response)
-        status = self._get_body_status(response)
-        if status:
-            self.status = status
+        self._raise_if_bad_http_status_and_method(response)
+
+        if self._is_empty(response):
+            self.resource = None
+        else:
+            try:
+                self.resource = self.get_outputs(response)
+            except DeserializationError:
+                self.resource = None
+
+        self.set_async_url_if_present(response)
+
+        if response.status_code in {200, 201, 202, 204}:
+            self.initial_status_code = response.status_code
+            if self.async_url or self.location_url or response.status_code == 202:
+                self.status = 'InProgress'
+            elif response.status_code == 201:
+                status = self._get_provisioning_state(response)
+                self.status = status or 'InProgress'
+            elif response.status_code == 200:
+                status = self._get_provisioning_state(response)
+                self.status = status or 'Succeeded'
+            elif response.status_code == 204:
+                self.status = 'Succeeded'
+                self.resource = None
```

Comment:
Would it be worth raising an exception if no clause is executed? That way if the acceptable status codes changes you won't forget to add a clause for it.